### DISCUSSION TITLE
Project admin, milestones, and roadmap

### DISCUSSION THREAD

**Discussion by christian-oreilly** [niivue/ipyniivue#250] [General]
This "issue" is more of an administrative log to help us manage the initial development of this project. We can close it and move this to other more appropriate management tools down the line. For now, I suggest the following target for ipyniivue-experimental based on the AnyWidget approach:

Develop a widget allowing us to replicate, in JupyterNotebook, with Python code, the demo already prepared for NiiVue. The functionalities must be usable from the Python side (i.e., we don't want just to be able to display JS demos in Jupyter Notebooks; we want to be able to implement them in Python using the JS wrapper). I suggest we implement these are different notebooks and cross them off the following list when done:

- [x] [Basic multiplanar](https://niivue.github.io/niivue/demos/features/basic.multiplanar.html)
- [x] [Sync mesh](https://niivue.github.io/niivue/demos/features/sync.mesh.html)
- [x] [Bidirectional Sync](https://niivue.github.io/niivue/demos/features/sync.bidirectional.html)
- [x] [Color maps for voxels](https://niivue.github.io/niivue/demos/features/colormaps.html) and [meshes](https://niivue.github.io/niivue/demos/features/colormaps.mesh.html)
- [x] [Background masks overlays](https://niivue.github.io/niivue/demos/features/mask.html)
- [x] [Alpha and asymmetric statistical thresholds](https://niivue.github.io/niivue/demos/features/alphathreshold.html)
- [x] [Test images](https://niivue.github.io/niivue/demos/features/test_images.html)
- [x] [Drag and drop](https://niivue.github.io/niivue/demos/features/draganddrop.html)
- [ ] [Select font](https://niivue.github.io/niivue/demos/features/selectfont.html)
- [ ] [Connectome](https://niivue.github.io/niivue/demos/features/connectome.html) [Connectome API](https://niivue.github.io/niivue/demos/features/connectome.api.html)
- [ ] [Minimal user interface with menus](https://niivue.github.io/niivue/demos/features/ui.html) see [niivue-ui for a more advanced user interface](https://github.com/niivue/niivue-ui)
- [ ] [Meshes (GIfTI, FreeSurfer, MZ3, OBJ, STL, legacy VTK)](https://niivue.github.io/niivue/demos/features/meshes.html)
- [x] [Mesh MatCaps](https://niivue.github.io/niivue/demos/features/mesh.matcap.html)
- [ ] [Mesh Statistics](https://niivue.github.io/niivue/demos/features/mesh.stats.html)
- [x] [Mesh layers (GIfTI, FreeSurfer, MZ3)](https://niivue.github.io/niivue/demos/features/mesh.layers.html)
- [ ] [Load Mesh Layers](https://niivue.github.io/niivue/demos/features/mesh.loader.html)
- [x] [4D mesh time series (GIfTI, FreeSurfer, MZ3)](https://niivue.github.io/niivue/demos/features/mesh.4D.html)
- [ ] [annot mesh atlases](https://niivue.github.io/niivue/demos/features/mesh.atlas.html) and [GIFTI mesh atlases](https://niivue.github.io/niivue/demos/features/mesh.atlas.gii.html) and [MGH mesh atlases](https://niivue.github.io/niivue/demos/features/mesh.atlas.mgh.html)
- [ ] [GIfTI Meshes with NIfTI2 Curvature](https://niivue.github.io/niivue/demos/features/mesh.curv.html)
- [x] Anti-Aliasing: faster (off) or better (on)
- [x] Auto-select [Mesh](https://niivue.github.io/niivue/demos/features/mesh.freesurfer.html)   [Voxel](https://niivue.github.io/niivue/demos/features/vox.aaAUTO.html)
- [x] OFF (faster) [Mesh](https://niivue.github.io/niivue/demos/features/mesh.freesurfer.aaOFF.html)   [Voxel](https://niivue.github.io/niivue/demos/features/vox.aaOFF.html)
- [x] ON (better) [Mesh](https://niivue.github.io/niivue/demos/features/mesh.freesurfer.aaON.html)   [Voxel](https://niivue.github.io/niivue/demos/features/vox.aaON.html)
- [ ] [FreeSurfer mask editing](https://niivue.github.io/niivue/demos/features/freesurfer.html)
- [ ] [FreeSurfer point sets](https://niivue.github.io/niivue/demos/features/pointset.html)
- [ ] [4D mesh time series (CIFTI-2)](https://niivue.github.io/niivue/demos/features/cifti.4D.html)
- [x] [Tractography (TCK, TRK, TRX, VTK)](https://niivue.github.io/niivue/demos/features/tracts.html)
- [ ] [Tractography groups (TRX)](https://niivue.github.io/niivue/demos/features/tracts.group.html)
- [ ] [Advanced tractography (TCK, TRK, TRX, VTK)](https://niivue.github.io/niivue/demos/features/tracts2.html)
- [ ] [4D Time series data (fMRI, DTI, ASL, etc) using thumbnail for rapid loading](https://niivue.github.io/niivue/demos/features/timeseries.html)
- [ ] [4D Time series data (fMRI, DTI, ASL, etc) initially showing only the first volumes for rapid loading](https://niivue.github.io/niivue/demos/features/timeseries2.html)
- [ ] [AFNI volumes](https://niivue.github.io/niivue/demos/features/afni.html)
- [ ] [BrainVoyager meshes](https://niivue.github.io/niivue/demos/features/brainvoyager.html)
- [ ] [x3d meshes](https://niivue.github.io/niivue/demos/features/x3d.html)
- [ ] [Clip planes](https://niivue.github.io/niivue/demos/features/clipplanes.html)
- [x] [Drawing](https://niivue.github.io/niivue/demos/features/draw2.html)
- [x] [Drawing user interface](https://niivue.github.io/niivue/demos/features/draw.ui.html)
- [ ] [Torso regions](https://niivue.github.io/niivue/demos/features/torso.html)
- [ ] [Carotid Artery-Computed Tomographic Angiography Scoring](https://niivue.github.io/niivue/demos/features/cactus.html)
- [x] [Voxel atlas](https://niivue.github.io/niivue/demos/features/atlas.html)
- [x] [Sparse voxel atlas](https://niivue.github.io/niivue/demos/features/atlas.sparse.html)
- [ ] [Modulation](https://niivue.github.io/niivue/demos/features/modulate.html)
- [ ] [Multiplanar Layout (Auto, Column, Grid, Row)](https://niivue.github.io/niivue/demos/features/layout.html)
- [x] [Modulation of scalar volumes](https://niivue.github.io/niivue/demos/features/modulateScalar.html)
- [ ] [Modulation of AFNI statistics](https://niivue.github.io/niivue/demos/features/modulateAfni.html)
- [ ] [Fixed canvas size](https://niivue.github.io/niivue/demos/features/fixedsize.html)
- [ ] [Segmented image with named labels](https://niivue.github.io/niivue/demos/features/segment.html)
- [x] [Mosaics](https://niivue.github.io/niivue/demos/features/mosaics.html)
- [ ] [Advanced mosaics](https://niivue.github.io/niivue/demos/features/mosaics2.html)
- [ ] [World space](https://niivue.github.io/niivue/demos/features/worldspace.html)
- [ ] [Advanced world space](https://niivue.github.io/niivue/demos/features/worldspace2.html)
- [ ] [Images with shear](https://niivue.github.io/niivue/demos/features/shear.html)
- [ ] [Multiuser basic 3D](https://niivue.github.io/niivue/demos/features/multiuser.3d.html)
- [ ] [Multiuser Time Series](https://niivue.github.io/niivue/demos/features/multiuser.timeseries.html)
- [ ] [Multiuser color maps](https://niivue.github.io/niivue/demos/features/multiuser.imageoptions.html)
- [ ] [Multiuser meshes](https://niivue.github.io/niivue/demos/features/multiuser.meshes.html)
~~- [ ] [User scripting](https://niivue.github.io/niivue/demos/features/scripts.html)~~ N/A? Besides, `eval` not safe to have in code
- [ ] [Shiny volume rendering](https://niivue.github.io/niivue/demos/features/shiny.volumes.html)
- [x] [Dragging callbacks](https://niivue.github.io/niivue/demos/features/dragCallback.html)
- [x] [Voxels with complex numbers](https://niivue.github.io/niivue/demos/features/complex.html)
- [ ] Additive [voxels](https://niivue.github.io/niivue/demos/features/additive.voxels.html) and [mesh](https://niivue.github.io/niivue/demos/features/additive.mesh.html) as well as [negative colormaps](https://niivue.github.io/niivue/demos/features/additive.mesh.negative.html)
- [ ] [DICOM Manifest](https://niivue.github.io/niivue/demos/features/dicom.manifest.html)
- [x] [Load a document](https://niivue.github.io/niivue/demos/features/document.load.html)
- [ ] [Download a document with 3D render](https://niivue.github.io/niivue/demos/features/document.3d.html)
- [ ] [Download a document with drawing](https://niivue.github.io/niivue/demos/features/document.drawing.html)
- [ ] [Download a document with meshes](https://niivue.github.io/niivue/demos/features/document.meshes.html)
- [x] [Save a scene as HTML](https://niivue.github.io/niivue/demos/features/save.html.html)
- [ ] [Save a scene with custom HTML](https://niivue.github.io/niivue/demos/features/save.custom.html.html)
- [ ] [Show labels](https://niivue.github.io/niivue/demos/features/labels.html)

Besides ensuring these functionalities (and by integrating as notebooks ran by the CI on PR, ensuring that there is no code regression on these functionalities), I think we need to target the following objectives:
- [x] Every public class, function, class method,  and class attribute has a doc string following pydocstyle
- [x] The code is PEP8 compliant the linting is automatically checked and enforced through CI on every PR 
- [x] The documentation is automatically generated and pushed to a doc website (see hooks implemented for ipyniivue)
- [ ] Test coverage above 90%

Final objective:
- [ ] Publication of a paper on this package.

**Comment by christian-oreilly** [comment]
@bcalford I think we can check out the Drawing examples from that list, right? Please go ahead and check any other examples from the list above that you think we are already covering.

**Comment by christian-oreilly** [comment]
Hey @AnthonyAndroulakis I just wanted to check with you: Do you keep the status of this list up-to-date in general? Do you have the NiiVue meetings on some calendar? As I am sure  you remember, they are biweekly. We had one today. Could you make sure that you update this list (i.e., check whatever is finished) before these meeting? That way we can use this when the group wants an update and you are not there.

**Comment by AnthonyAndroulakis** [comment]
This list is up to date currently, yes, and I update it whenever I work on the project. I've been unable to join the past few meetings due to work conflicts, but I'll try to make more updates to the ipyniivue project.

**Comment by AnthonyAndroulakis** [comment]
have not been updating this list lately..but going to move it to discussions to leave the issues section to only contain issues
